pkg/sql/util/index_util.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2021 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/pb/plan"
+	"github.com/matrixorigin/matrixone/pkg/sql/plan/function/builtin/multi"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+	"strconv"
+)
+
+// unique + index number + index name
+var prefixUniqueIndexColName string = "__mo_"
+
+func BuildIndexTableName(unique bool, indexNumber int, indexName string) string {
+	var name string
+	name = prefixPriColName
+	if unique {
+		name += "unique_"
+	} else {
+		name += "secondary_"
+	}
+	name += strconv.Itoa(indexNumber)
+	name += "_"
+	name += indexName
+
+	return name
+}
+
+func BuildUniqueKeyBatch(vecs []*vector.Vector, Attrs []string, p []*plan.ColDef, proc *process.Process) (*batch.Batch, int) {
+	b := &batch.Batch{
+		// make sure that when batch is cleaned, origin batch is not affected
+		Attrs: make([]string, 0),
+		Vecs:  make([]*vector.Vector, 0),
+	}
+	if p[0].IsCPkey {
+		names := SplitCompositePrimaryKeyColumnName(p[0].Name)
+		cPkeyVecMap := make(map[string]*vector.Vector)
+		for num, attrName := range Attrs {
+			for _, name := range names {
+				if attrName == name {
+					cPkeyVecMap[name] = vecs[num]
+				}
+			}
+		}
+		vs := make([]*vector.Vector, 0)
+		for _, name := range names {
+			v := cPkeyVecMap[name]
+			vs = append(vs, v)
+		}
+		vec, _ := multi.Serial(vs, proc)
+		b.Attrs = append(b.Attrs, p[0].Name)
+		b.Vecs = append(b.Vecs, vec)
+	} else {
+		for i, name := range Attrs {
+			if p[0].Name == name {
+				b.Attrs = append(b.Attrs, name)
+				b.Vecs = append(b.Vecs, vecs[i])
+			}
+		}
+	}
+	v, rowNum, needClean := compactUniqueKeyBatch(b.Vecs[0], proc, p[0].IsCPkey)
+	if needClean {
+		b.Cnt = 1
+		b.Vecs[0] = v
+	} else {
+		b.Cnt = 2
+		b.Vecs[0] = v
+	}
+	return b, rowNum
+}
+
+func compactUniqueKeyBatch(v *vector.Vector, proc *process.Process, isCPkey bool) (*vector.Vector, int, bool) {
+	var rowNum int
```

Comment:
I don't see how this function will be useful.   

Also, the NewWithFixed stuff, now you cannot use a vector from make().   It will corrupt memory.
And this method is horribly inefficient, you will keep on reallocate ns.

Reply:
yeah, I will think about it carelly and begin a meet